distributed/scheduler.py

```diff
@@ -8205,6 +8280,127 @@ def _worker_full(ws: WorkerState, saturation_factor: float) -> bool:
     return _task_slots_available(ws, saturation_factor) <= 0
 
 
+def _previous_in_linear_chain(ts: TaskState, cutoff: int) -> TaskState | None:
+    if len(ts.dependents) != 1 or len(ts.dependencies) > cutoff:
+        return None
+
+    prev: TaskState | None = None
+    for dts in ts.dependencies:
+        if len(dts.dependents) > cutoff:  # widely-shared; ignore it
+            continue
+        if prev:
+            return None
+        prev = dts
+
+    return prev
+
+
+def _next_in_linear_chain(ts: TaskState, cutoff: int) -> TaskState | None:
+    if len(ts.dependents) != 1 or len(ts.dependencies) > cutoff:
```

Comment:
Why do we not walk tasks with widely shared *dependencies*? Why wouldn't we want to walk the chain

```mermaid
flowchart BT

A1 --> B
A2 --> B
A3 --> B
A4 --> B
A5 --> B
B --> C
C --> D
```

Reply:
Can we re-use / factor out some code from `fuse / fuse_linear`? IIUC the code there handles more complicated situations that'd also be interesting for us

Reply:
> I think len(ts.dependencies) > cutoff is just about setting an upper bound for iterating through dependencies, to constrain runtime.

I don't mind introducing a cutoff here to protect from unreasonable traversals but I'm struggling a lot with motivating this cut to be correlated to the number of workers in this case